### PR TITLE
Route native loads through torch_checkpointing

### PR DESCRIPTION
**Pull request by ivy-zhou**
Summary:
`TorchCheckpointingManager` could save but not load: `_load` raised
`NotImplementedError`, so a run configured for this backend could write
checkpoints and never resume from them. Implement the native load path,
completing the round trip.

Resolution mirrors the DCP manager. With no explicit step, the latest valid
step in the checkpoint folder is used; with an explicit step, a missing folder
or missing checkpoint is an error rather than a silent fresh start. When no
step is found, `initial_load_path` is used if configured, and otherwise the run
starts fresh. Step 0 is treated as a seed checkpoint and loads model state
only.

Step discovery only accepts directories that actually contain the backend's
`metadata.pkl`. A crashed save leaves a `step-N` directory with no metadata, and
without this check that partial directory would win the `max()` and be loaded
as the resume point. Directory listing, existence checks, and the load itself
all go through the configured `Storage`, so remote checkpoint folders work the
same as local ones.

The backend returns a plain dict rather than writing through the live objects,
so `_restore_state_dict` puts values back: `Stateful` entries get
`load_state_dict`, plain dicts are updated in place so callers holding a
reference see the result, and anything else that is not already the same object
raises rather than being silently skipped. Keys in `exclude_from_loading` are
dropped before the load, and an excluded key that is not in the state dict is
an error, matching the DCP manager.

Hugging Face loads are explicitly rejected for now; they land separately.

Ported from an internal change. Besides the path and naming translation, the
internal version defines `load()` with its own `if not self.enable` guard; here
the body moves to the `_load` hook and the guard is dropped, since
`BaseCheckpointManager` owns it.

Test Plan:
`pytest tests/unit_tests/test_torch_checkpointing.py`: 24 passed.

Four new tests: an explicit-step load restores both model weights and optimizer
state from the backend payload and passes the full state dict as the load
target; a latest-step load of `step-0` requests model state only; discovery and
load go through a configured non-local `Storage` rather than the filesystem;
and a folder containing valid `step-2` and `step-5`, a metadata-less `step-8`,
and a `tmp_step-9` resolves to `step-5`, so an interrupted save is not mistaken
for the newest checkpoint.

`pytest tests/unit_tests/test_checkpoint.py tests/unit_tests/observability/
torchtitan/experiments/torchft/tests/test_torchft_checkpoint.py` alongside the
above: 147 passed, 2 subtests passed.

Also verified:
- `TorchCheckpointingManager.__abstractmethods__` is empty; with `_load`
  implemented the manager now satisfies the full `BaseCheckpointManager`
  contract.
- `METADATA_FILE_NAME` and `CheckpointManager.load(checkpoint_id, into=...)`
  exist in `torch_checkpointing` 0.1.0.
- `ufmt` and `flake8 --config=.flake8` clean on both changed files.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/torchtitan/pull/4191).
* #4457
* #4277
* #4279
* #4278
* __->__ #4191

### CODE REVIEW

File: torchtitan/components/checkpointer/torch_checkpointing.py [hunk 1]
```diff
@@ -192,6 +194,36 @@ def _item_specs() -> dict[str, ItemSpec]:
     }
 
 
+def _restore_state_dict(
```
Comment (meetv18) [inline torchtitan/components/checkpointer/torch_checkpointing.py:149]: is this logic commonly shared with dcpv1?

**Review by meetv18 — APPROVED** [review]

**Review by fegin — COMMENTED** [review]
Only one comment, any reason we cannot make `_load` in the base class? What is the difference? Same question to `_states_to_load`.

File: torchtitan/components/checkpointer/torch_checkpointing.py (hunk 1 above)
Comment (ivy-zhou) [reply torchtitan/components/checkpointer/torch_checkpointing.py]: You're right, DCPv1 does do some version of this because the model's state needs to be explicitly reloaded because we flatten model state with a ModelWrapper class, but this was overly complicated, I fixed this and dropped the helper.

File: torchtitan/components/checkpointer/torch_checkpointing.py [hunk 2]
```diff
@@ -365,6 +398,11 @@ def _save(self, curr_step: int, last_step: bool = False) -> bool:
 
         return True
 
+    def _is_resumable_checkpoint(self, checkpoint_dir: str) -> bool:
+        return self._storage.isfile(
+            filesystem.join(checkpoint_dir, TORCH_CHECKPOINTING_METADATA_FILE_NAME)
+        )
```
Comment (fegin) [inline torchtitan/components/checkpointer/torch_checkpointing.py:410]: Let's add a docstring to mention the difference between `_is_resumable_checkpoint` and `_is_valid_checkpoint`.

**Review by fegin — APPROVED** [review]

**Comment by fegin** [comment]
Please add the docstring before merging.
